### PR TITLE
Initial package loader implementation, including VCS support for git

### PR DESCRIPTION
**Pull request by josephschorr**

### CODE REVIEW

File: vcs/checkouttool/checkouttool.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+			}
+
+			fmt.Printf("Performing checkout for path: %v", path)
+			fmt.Println()
+
+			pkgCacheDirectory := ".pkg"
+			packagePath, err, warning := vcs.PerformVCSCheckout(path, pkgCacheDirectory)
+
+			fmt.Printf("Path: %s", packagePath)
+			fmt.Println()
```
Comment (jzelinskie) [inline vcs/checkouttool/checkouttool.go]: what? just add a `\n` to the previous printf

File: vcs/metadiscovery.go [hunk 2]
```diff
@@ -1,5 +1,7 @@
 package vcs
 
+// Note: Discovery protocol is documented loosely here: https://golang.org/cmd/go/#hdr-Remote_import_paths
```
Comment (jzelinskie) [inline vcs/metadiscovery.go]: you should probably put this in the package docs

File: vcs/vcspath_test.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+	assertParses(t, "github.com/some/project@sometag", "github.com/some/project", "", "sometag", "")
+	assertParses(t, "github.com/some/project//somesubdir", "github.com/some/project", "", "", "somesubdir")
+	assertParses(t, "github.com/some/project//somesubdir:somebranch", "github.com/some/project", "somebranch", "", "somesubdir")
+	assertParses(t, "github.com/some/project//somesubdir@sometag", "github.com/some/project", "", "sometag", "somesubdir")
+
+	assertParseError(t, "")
+	assertParseError(t, "github.com/foo/bar@@blah")
+	assertParseError(t, "github.com/foo/bar:@blah")
+	assertParseError(t, "github.com/foo/../bar")
+}
```
Comment (jzelinskie) [inline vcs/vcspath_test.go:20]: This could be made table-driven.

File: vcs/checkouttool/checkouttool.go (hunk 1 above)
Comment (josephschorr) [inline vcs/checkouttool/checkouttool.go]: Done

File: vcs/metadiscovery.go (hunk 2 above)
Comment (josephschorr) [inline vcs/metadiscovery.go]: Done

File: vcs/vcspath_test.go (hunk 3 above)
Comment (josephschorr) [inline vcs/vcspath_test.go:20]: Think it is worth changing?

Comment (jzelinskie) [inline vcs/vcspath_test.go:20]: Only if you're consistently going to have table driven tests for everything -- basically pick a test style and stick with it.

Comment (josephschorr) [inline vcs/vcspath_test.go:20]: Okay, will do so

**Comment by josephschorr** [comment]
@jzelinskie: Tests changed to be table drive. Let me know if I missed any.

File: vcs/vcspath_test.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+type pathSuccessTest struct {
+	path           string
+	url            string
+	branchOrCommit string
+	tag            string
+	subpackage     string
+}
+
+var successTests = []pathSuccessTest{
+	pathSuccessTest{"github.com/some/project", "github.com/some/project", "", "", ""},
```
Comment (jzelinskie) [inline vcs/vcspath_test.go:23]: you can actually use a literal and forgo defining the type, if you'd like

``` go
var successTests = []struct{string,string,string,string,string}{
  {"github.com/some/project", "github.com/some/project", "", "", ""},
  ...
}
```

The type might make things more clear, tho. Use the above at your discretion.

Comment (josephschorr) [inline vcs/vcspath_test.go:23]: Oh, interesting. I think the type does make it a bit more clear, but definitely good to know for smaller types.

**Comment by josephschorr** [comment]
Will merge into master

**Pull request closed without merging**